src/Engine/Renderer/Mesh/Mesh.cpp

```diff
@@ -110,14 +107,32 @@ void Mesh::loadGeometry( const Core::Vector3Array& vertices, const std::vector<u
 }
 
 void Mesh::addData( const Vec3Data& type, const Core::Vector3Array& data ) {
-    m_v3Data[static_cast<uint>( type )] = data;
-    m_dataDirty[MAX_MESH + static_cast<uint>( type )] = true;
+    const int index = static_cast<uint>( type );
+    if ( !m_v3DataHandle[index].isValid() )
+    {
+        m_mesh.attribManager().addAttrib( m_v3DataHandle[index],
+                                          std::to_string( MAX_MESH + index ) );
+    }
+    m_mesh.attribManager().getAttrib( m_v3DataHandle[index] ).data() = data;
+    m_dataDirty[MAX_MESH + index] = true;
     m_isDirty = true;
 }
 
 void Mesh::addData( const Vec4Data& type, const Core::Vector4Array& data ) {
-    m_v4Data[static_cast<uint>( type )] = data;
-    m_dataDirty[MAX_MESH + MAX_VEC3 + static_cast<uint>( type )] = true;
+    const int index = static_cast<uint>( type );
+
+    /// to handle dummy set, if not valid, and dummy data, do nothing
+    if ( data.size() == 0 && !m_v4DataHandle[index].isValid() )

```

Comment:
Why is this only for Vec4Data ?

Reply:
now we have clarified handle and invalid data, here we should clean this hack